File: csharp/Wjybxx.Commons.Concurrent/src/Concurrent/EventLoopBuilder.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using Wjybxx.Disruptor;

namespace Wjybxx.Commons.Concurrent
{
/// <summary>
/// 
/// </summary>
public abstract class EventLoopBuilder<T> where T : IAgentEvent
{
    private IEventLoopGroup? _parent;
    private int index = -1;
    private RejectedExecutionHandler _rejectedExecutionHandler = RejectedExecutionHandlers.ABORT;
    private ThreadFactory? _threadFactory;

    private IEventLoopAgent<T>? _agent = EmptyAgent<T>.Inst;
    private IEventLoopModule? _mainModule;
    private int _batchSize = 1024;

    public EventLoopBuilder() {
    }

    public abstract IEventLoop Build();

    public IEventLoopGroup? Parent {
        get => _parent;
        set => _parent = value;
    }

    /// <summary>
    /// Parent为当前EventLoop分配的索引
    /// </summary>
    public int Index {
        get => index;
        set => index = value;
    }

    public RejectedExecutionHandler RejectedExecutionHandler {
        get => _rejectedExecutionHandler;
        set => _rejectedExecutionHandler = value ?? throw new ArgumentNullException(nameof(value));
    }

    /// <summary>
    /// 线程工厂
    /// </summary>
    public ThreadFactory? ThreadFactory {
        get => _threadFactory;
        set => _threadFactory = value;
    }

    /// <summary>
    /// 事件循环的内部代理
    /// </summary>
    public IEventLoopAgent<T>? Agent {
        get => _agent;
        set => _agent = value;
    }

    /// <summary>
    /// 事件循环的主模块
    /// </summary>
    public IEventLoopModule? MainModule {
        get => _mainModule;
        set => _mainModule = value;
    }

    /// <summary>
    /// 最多连续处理多少个事件必须执行一次Update
    /// </summary>
    public int BatchSize {
        get => _batchSize;
        set => _batchSize = value;
    }
}

public class DisruptorEventLoopBuilder<T> : EventLoopBuilder<T> where T : IAgentEvent
{
    private EventSequencer<T> eventSequencer;
    private WaitStrategy? waitStrategy;
    private bool cleanEventAfterConsumed = true;
    private bool cleanBufferOnExit = true;

    private void CheckBuild() {
        if (ThreadFactory == null) {
            ThreadFactory = new DefaultThreadFactory("DisruptorEventLoop");
        }
        if (eventSequencer == null) {
            throw new IllegalStateException("eventSequencer is null");
        }
    }

#if UNITY_EDITOR
    public override IEventLoop Build() {
        CheckBuild();
        return new DisruptorEventLoop<T>(this);
    }
#else
    public override DisruptorEventLoop<T> Build() {
        CheckBuild();
        return new DisruptorEventLoop<T>(this);
    }
#endif

    /// <summary>
    /// 事件序列生成器
    /// 注意：应当避免使用无超时的等待策略，EventLoop需要处理定时任务，不能一直等待生产者。
    /// </summary>
    /// <exception cref="ArgumentNullException"></exception>
    public EventSequencer<T>? EventSequencer {
        get => eventSequencer;
        set => eventSequencer = value ?? throw new ArgumentNullException(nameof(value));
    }

    /// <summary>
    /// 等待策略
    /// 1.如果未显式指定，则使用<see cref="Sequencer.WaitStrategy"/>中的默认等待策略。
    /// 2.应当避免使用无超时的等待策略，EventLoop需要处理定时任务，不能一直等待生产者。
    /// </summary>
    /// <exception cref="ArgumentNullException"></exception>
    public WaitStrategy? WaitStrategy {
        get => waitStrategy;
        set => waitStrategy = value;
    }

    /// <summary>
    /// 在消费事件后是否调用<see cref="IAgentEvent.Clean()"/>方法清理引用数据 
    /// </summary>
    public bool CleanEventAfterConsumed {
        get => cleanEventAfterConsumed;
        set => cleanEventAfterConsumed = value;
    }

    /// <summary>
    /// EventLoop在退出的时候是否清理buffer
    /// 1. 默认清理
    /// 2. 如果该值为true，意味着当前消费者是消费者的末端，或仅有该EventLoop消费者。
    /// </summary>
    public bool CleanBufferOnExit {
        get => cleanBufferOnExit;
        set => cleanBufferOnExit = value;
    }
}
}

File: java/Commons-Concurrent/src/main/java/cn/wjybxx/concurrent/EventLoopBuilder.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.concurrent;


import cn.wjybxx.base.Preconditions;
import cn.wjybxx.disruptor.EventSequencer;
import cn.wjybxx.disruptor.Sequencer;
import cn.wjybxx.disruptor.WaitStrategy;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ThreadFactory;

/**
 * @param <T> 内部事件类型
 * @author wjybxx
 * date 2023/4/11
 */
public abstract class EventLoopBuilder<T extends IAgentEvent> {

    private IEventLoopGroup parent;
    private int index = -1;
    private RejectedExecutionHandler rejectedExecutionHandler = RejectedExecutionHandlers.abort();
    private ThreadFactory threadFactory;

    private long consumerId;
    private IEventLoopAgent<T> agent;
    private final List<EventLoopModule> moduleList = new ArrayList<>();
    private int batchSize = 1024;

    public abstract IEventLoop build();

    public IEventLoopGroup getParent() {
        return parent;
    }

    public EventLoopBuilder<T> setParent(IEventLoopGroup parent) {
        this.parent = parent;
        return this;
    }

    /** Parent为当前EventLoop分配的索引 */
    public int getIndex() {
        return index;
    }

    public EventLoopBuilder<T> setIndex(int index) {
        this.index = index;
        return this;
    }

    /** 线程工厂 */
    public ThreadFactory getThreadFactory() {
        return threadFactory;
    }

    public EventLoopBuilder<T> setThreadFactory(ThreadFactory threadFactory) {
        this.threadFactory = threadFactory;
        return this;
    }

    /** 拒绝策略 */
    public RejectedExecutionHandler getRejectedExecutionHandler() {
        return rejectedExecutionHandler;
    }

    public EventLoopBuilder<T> setRejectedExecutionHandler(RejectedExecutionHandler rejectedExecutionHandler) {
        this.rejectedExecutionHandler = rejectedExecutionHandler;
        return this;
    }

    /** 事件循环的消费者id - 未指定将使用{@link Thread#threadId()} */
    public long getConsumerId() {
        return consumerId;
    }

    public EventLoopBuilder<T> setConsumerId(long consumerId) {
        this.consumerId = consumerId;
        return this;
    }

    /** 事件循环的内部代理 */
    public IEventLoopAgent<T> getAgent() {
        return agent;
    }

    public EventLoopBuilder<T> setAgent(IEventLoopAgent<T> agent) {
        this.agent = agent;
        return this;
    }

    /** 事件循环的模块 */
    public List<EventLoopModule> getModules() {
        return moduleList;
    }

    /** 添加模块 */
    public EventLoopBuilder<T> addModule(EventLoopModule module) {
        Objects.requireNonNull(module);
        moduleList.add(module);
        return this;
    }

    public EventLoopBuilder<T> addModules(List<EventLoopModule> modules) {
        Preconditions.checkNullElements(modules);
        moduleList.addAll(modules);
        return this;
    }

    /**
     * 每次最多处理多少个事件就尝试执行一次主循环方法
     * 该值越小：线程间的同步开销越多；越不容易阻塞生产者（有界Buffer）；EventLoop更容易响应取消；
     * 该值越大：消费者的吞吐量越好，生产者的吞吐量则会降低（有界Buffer）；EventLoop对关闭信号的响应越慢。
     */
    public int getBatchSize() {
        return batchSize;
    }

    public EventLoopBuilder<T> setBatchSize(int batchSize) {
        this.batchSize = batchSize;
        return this;
    }
    //

    public static <T extends IAgentEvent> DisruptorBuilder<T> newDisruptBuilder() {
        return new DisruptorBuilder<>();
    }

    public static <T extends IAgentEvent> DisruptorBuilder<T> newDisruptBuilder(EventSequencer<? extends T> eventSequencer) {
        return new DisruptorBuilder<T>()
                .setEventSequencer(eventSequencer);
    }

    //

    public static class DisruptorBuilder<T extends IAgentEvent> extends EventLoopBuilder<T> {

        private EventSequencer<? extends T> eventSequencer;
        private WaitStrategy waitStrategy;

        //

        @Override
        public DisruptorBuilder<T> setParent(IEventLoopGroup parent) {
            super.setParent(parent);
            return this;
        }

        @Override
        public DisruptorBuilder<T> setIndex(int index) {
            super.setIndex(index);
            return this;
        }

        @Override
        public DisruptorBuilder<T> setRejectedExecutionHandler(RejectedExecutionHandler rejectedExecutionHandler) {
            super.setRejectedExecutionHandler(rejectedExecutionHandler);
            return this;
        }

        @Override
        public DisruptorBuilder<T> setThreadFactory(ThreadFactory threadFactory) {
            super.setThreadFactory(threadFactory);
            return this;
        }

        @Override
        public DisruptorBuilder<T> setConsumerId(long consumerId) {
            super.setConsumerId(consumerId);
            return this;
        }

        @Override
        public DisruptorBuilder<T> setAgent(IEventLoopAgent<T> agent) {
            super.setAgent(agent);
            return this;
        }

        @Override
        public DisruptorBuilder<T> addModule(EventLoopModule module) {
            super.addModule(module);
            return this;
        }

        @Override
        public DisruptorBuilder<T> addModules(List<EventLoopModule> modules) {
            super.addModules(modules);
            return this;
        }

        public DisruptorBuilder<T> setBatchSize(int batchSize) {
            super.setBatchSize(batchSize);
            return this;
        }

        @Override
        public DisruptorEventLoop<T> build() {
            if (getThreadFactory() == null) {
                setThreadFactory(new DefaultThreadFactory("DisruptorEventLoop"));
            }
            if (eventSequencer == null) {
                throw new IllegalStateException("eventSequencer is null");
            }
            return new DisruptorEventLoop<>(this);
        }

        //

        /**
         * 事件序列生成器
         * 注意：应当避免使用无超时的等待策略，EventLoop需要处理定时任务，不能一直等待生产者。
         */
        public EventSequencer<? extends T> getEventSequencer() {
            return eventSequencer;
        }

        public DisruptorBuilder<T> setEventSequencer(EventSequencer<? extends T> eventSequencer) {
            this.eventSequencer = eventSequencer;
            return this;
        }

        /**
         * 等待策略
         * 1.如果未显式指定，则使用{@link Sequencer#getWaitStrategy()}中的默认等待策略。
         * 2.应当避免使用无超时的等待策略，EventLoop需要处理定时任务，不能一直等待生产者。
         */
        public WaitStrategy getWaitStrategy() {
            return waitStrategy;
        }

        public DisruptorBuilder<T> setWaitStrategy(WaitStrategy waitStrategy) {
            this.waitStrategy = waitStrategy;
            return this;
        }
    }

}